### PR TITLE
Sockets - Nara

### PR DESCRIPTION
**Pull request by bonara**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | It was challenging to manage dependencies between my classes and to decide which method should belong to a class. I had an issue deciding where to include the list_all_rooms method. Since list_all_rooms method creates 20 rooms, I need to make some changes to handle hotels with different room numbers in the future. 
What was a design decision you made that changed over time over the project? | First, I didn't think having a separate class for DateRange mattered. However, the more I thought about dependencies and methods that will be used in other classes, it made sense to have it as a separate class. 
What was a concept you gained clarity on, or a learning that you'd like to share? | I learned that it is hard to make changes to tightly coupled classes. Changing only one keyword argument can create many problems in other classes.
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | An example of a nominal test that I wrote is checking that the reservation total cost is calculated correctly. It is nominal because we are checking for the expected behavior. 
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | Testing for a case when a user tries to reserve an already booked block room is an example of an edge case. It is an edge case because the room is already reserved and user shouldn't be able to double book a room. 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | In most of my codes, I haven't written pseudocode. However, I outlined the methods and added requirements to get a clear understanding of the user case. I am still trying to get in a habit of writing tests first. I have been using pry to test my methods.

### CODE REVIEW

File: lib/block.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+module Hotel
+  class Block
+    attr_accessor :start_date, :end_date, :date_range, :rooms, :discounted_rate, :block_reservations
+    def initialize(start_date:, end_date:, rooms:, discounted_rate:)
+      @date_range = Hotel::DateRange.new(start_date, end_date)
+      @rooms = rooms
+      @discounted_rate = discounted_rate
+      @block_reservations = []
+      @rooms.each {|room| room.room_rate = discounted_rate}
```
Comment (dHelmgren) [inline lib/block.rb:13]: This is permanently changing the rate for a room, unless you set it back at some later time.

File: lib/date_range.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
+require 'date'
+
+module Hotel
+  class DateRange
+    attr_reader :start_date, :end_date
+    def initialize(start_date, end_date)
+      @start_date = Date.parse(start_date)
```
Comment (dHelmgren) [inline lib/date_range.rb:7]: You always have the option of forcing whoever is sending you the data to parse it themselves, just assuming properly formatted `Date` types in your code.

File: lib/hotel_dispatcher.rb [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    attr_accessor :rooms, :reservations, :blocks
+    def initialize
+      @rooms = Hotel::Room.list_all_rooms
+      @reservations = []
+      @blocks = Hash.new(0)
+    end
+
+    def reserve(start_date, end_date)
+      available_rooms = find_available_room(start_date, end_date)
+      raise ArgumentError.new("Room is not available") if available_rooms.empty?
```
Comment (dHelmgren) [inline lib/hotel_dispatcher.rb]: better way to phrase this error: "No rooms available from ${start_date} to ${end_date}".

File: lib/room.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+module Hotel
+  class Room
+    attr_accessor :room_num, :room_rate
+
```
Comment (dHelmgren) [inline lib/room.rb:4]: This class is pretty thin. Does its usefulness justify its existence?

File: lib/hotel_dispatcher.rb [hunk 5]
```diff
@@ -0,0 +1,138 @@
+require 'date'
+
+require_relative 'room'
+require_relative 'reservation'
+require_relative 'block'
+require_relative 'date_range'
+
+module Hotel
+  class HotelDispatcher
```
Comment (dHelmgren) [inline lib/hotel_dispatcher.rb]: Dispatcher isn't a great word for what this class does. Manager might be better. Dispatch implies that it's sending out hotel rooms, which doesn't quite fit.

File: lib/hotel_dispatcher.rb [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    def add_block(start_date, end_date, room_nums, discounted_rate, access_code)
+      rooms = room_nums.map do |room_num|
+        self.find_room(room_num)
+      end
+      
+      rooms.each do |room|
+        if list_reserved_rooms(start_date, end_date).include?(room)
+          raise ArgumentError.new("Room #{room.room_num} has already been reserved")
+        elsif list_block_rooms(start_date, end_date).include?(room)
+          raise ArgumentError.new("Room#{room.room_num} has already been added to a block")
```
Comment (dHelmgren) [inline lib/hotel_dispatcher.rb]: So with errors around the creation of an object, I would actually put these inside the class that is being created. It's the `Block`s responsibility to know what makes a valid `Block`, not the `HotelDispatcher`s.

File: lib/hotel_dispatcher.rb [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        rooms: rooms,
+        discounted_rate: discounted_rate
+      )
+      raise ArgumentError.new("Access code is already taken, please try with a new one.") if @blocks.keys.include?(access_code)
+      @blocks[access_code] = new_block
+    end
+
+    def reserve_from_block(room_num, access_code)
+      raise ArgumentError.new("Access code #{access_code} is invalid") if !(@blocks.keys.include?(access_code))
+      raise ArgumentError.new("This block room has already been reserved.") if list_block_reserved_rooms(access_code).include?(room_num)
```
Comment (dHelmgren) [inline lib/hotel_dispatcher.rb]: Again, most of these errors should belong to the `Block` class.

File: lib/hotel_dispatcher.rb [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+          reservation_by_date << r
+        end
+      end
+      return reservation_by_date
+    end 
+
+    def list_reserved_rooms(start_date, end_date)
+      reserved_rooms = [] 
+      @reservations.each do |r|
+        if r.date_range.is_overlapped?(start_date, end_date)
```
Comment (dHelmgren) [inline lib/hotel_dispatcher.rb]: Well done!

**Comment by dHelmgren** [comment]
# Hotel
## What We're Looking For


Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | Yes
Answer comprehension questions | yes
**Design** |
Each class is responsible for a single piece of the program | Mostly, your `HotelDispatch` class does a lot of work that should go in the classes that it manages.
Classes are loosely coupled | Mostly, but your coupling happens mostly in the manager.
**Wave 1** |
List rooms | Yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | Good!
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | Nice! Your tests are well constructed
**Wave 3** |
Create a block of rooms | Yes
Check if a block has rooms | yes
Reserve a room from a block | yes
Test coverage | Good, but your blocks permanently change how much a room costs, see comment.
**Fundamentals** |
Names variables, classes and modules appropriately | Mostly, see comments.
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | yes
Understands the differences between class and instance methods | yes
Appropriately uses iterators and `Enumerable` methods | yes
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a namespace | yes
**Wrap Up** |
There is a refactors.txt file | yes
The file provides a roadmap to future changes | yes
**Additional Feedback** | Good work on this. There are a few comments to help improve the code further, but the baseline you turned in here is very solid.
